### ISSUE TITLE
Generalize and enrich REST API error representations

### ISSUE THREAD

**Issue by timoreimann** [gambol99/go-marathon#101]
HTTP errors coming in from Marathon's REST API have a fairly limited representation in the library. In particular, the following two shortcomings exist:

1. HTTP error codes are not included, making fault analysis more difficult for clients than necessary.
1. For errors not represented by dedicated error structs, an attempt is made to retrieve a top-level `message` property from the HTTP response. Unfortunately, Marathon does not always provide such a property, causing go-marathon to fall back to a rather non-informative _unknown error_ description. Or Marathon provides properties in addition to `message` which are missed out. An example for the former is a 409 returned after a `POST /v2/apps/{appId}` (which also contains the deployment ID in a `deployments` array); a case for the latter is a 422 response (_Unprocessable Entity_) when an invalid app ID is passed to Marathon. (The resulting response will repeat the violating ID in an `attribute` property along with error details in an `error` property, both wrapped in an `errors` object.)

It'd be great if go-marathon could be more verbose regarding the error details it exposes.

**Comment by timoreimann** [comment]
I was already thinking about a solution approach and came up with the following:

1. Store the return code in a new base error struct.
2. Store the raw, unparsed JSON message in the error struct (and maybe add a `String()` implementation to print it nicely).

For errors where it makes sense that clients may want to react to it programmatically, we can continue to provide type-specific error structs and parse the JSON into individual fields. Personally, I yet have to find a case where I actually do something different than printing the error and let the client deal with it, or simply react to it on the sole basis of the response code. Thus, I'd argue that simply exposing the raw JSON string is more important.

Let me know what you think. If my approach is deemed worthy, I'd be more than happy to provide a PR.

**Comment by gambol99** [comment]
Hi @timoreimann ... the idea seems sounds to me, it's similar to another PR which was raised. I personally prefer library out abstract out the errors, but ErrItemNotFound, ErrPermissionDenied etc rather than handing back the specifics, but your wrapping the details in a base error, wrapped in a generic error type .. i.e MarathonGenericError{code, json_response} <- ErrItemNotFound and allowing the consumer to choose to just if they want to look into the specifics ... that's sounds good to me ..

**Comment by timoreimann** [comment]
@gambol99 I'm with you that, in general, strongly typed errors should be preferred over just dumping JSON strings. I think that once the library has matured enough with regards to error representation (that is, all kinds of Marathon errors are captured by well-populated, distinct error types and everything is well documented), we could remove the low level details again or keep them in only as a fallback when we have nothing better to offer. As we're mostly lacking typed error details for now, however, I think the proposed approach is a good interim one.

I will follow up with a PR as soon as I can. Thanks!

**Comment by eicca** [comment]
There's one more possible solution used by https://github.com/google/go-github: we can return 3 arguments result, response, error. In this case error can be very simple and response will contain all the logic. It will be also more flexible, for example in case we still want to check response even if error didn't occur. 

I'm not saying that it's better than proposed option, just want to be complete in our considerations :)

Unfortunately I haven't have a need to use different error types yet, so I can't really tell what I would prefer. When I was discussing previous similar PR I checked lots of different API wrapper libraries and only github library was really handling different errors. Most of the others just returned errors, not even library specific.

**Comment by timoreimann** [comment]
Interesting concept. I think it's a good idea to consider different approaches.

I digged a bit around the go-github code but couldn't really find out how they use the response. Do you have a better understanding than me?

The downside of the approach (apart from leaking abstractions) is that it increases the API surface area because you now have to deal with three parameters. Considering that we currently have no use for it other than passing through error messages, I'd personally opt to go with the error hierarchy approach.

**Comment by mattes** [comment]
@eicca Let's not expose the raw http data/ status codes to the user. I use this lib because I want to abstract the transportation layer away, meaning not dealing with raw marathon responses. :-)

I would like to propose a third approach that allows us to keep the `error` return type. `error` is just an [interface](https://golang.org/ref/spec#Errors). We can have better errors with minimal code changes. See an example: http://play.golang.org/p/t-Ff2Syvg_

**Comment by Raffo** [comment]
@timoreimann Did you have a look at this: https://github.com/Raffo/go-marathon/commit/bc23d5ae0467b23776c18b633c93f08473a78164 I was working on the same and had similar thoughts. Also, the PR @gambol99 @eicca are referring to is this one: https://github.com/gambol99/go-marathon/pull/77 . The work was indeed crappy, but you can see the idea and the discussion. I agree with you that we need something better here and I'm pushing hard to have this done with the community.

**Comment by timoreimann** [comment]
@mattes Using type assertions is basically what I had in mind as well for those cases where you want to programmatically access certain fields contained in the error or want to further process the error in the code. As I reasoned in my first two posts to this issue, however, the common case for me is to simply display Marathon API errors so users know what to do. That's why I'd like to have the HTTP error code and the raw JSON embedded in each error struct we define so that we can show that if nothing better is available.

Instead of embedding, we could restrict the raw error-reporting error struct to situations only where we do not have a higher level error instance available (which, for now, is the case for almost every error we report). However, I'd like at least the HTTP error response code to be available at all times as this will make debugging much easier.

Either way, my proposal should allow both proper abstraction as well as retrieving low-level details where needed. I hope it'll become clear as soon as I send out a PR.

@Raffo I saw the PR but as it got closed I didn't pay too close attention -- apologies. I believe we're on similar tracks, however.

**Comment by Raffo** [comment]
@timoreimann I believe that too, let's try to do one thing and do it good. Can we start from this? https://github.com/Raffo/go-marathon/commit/bc23d5ae0467b23776c18b633c93f08473a78164

**Comment by eicca** [comment]
I played around with the different errors implementations and I agree with all of you about having structs for different errors.

We need to try to implement it in a clean way with a good methods documentation, so user will know which error types a method could return.

> @timoreimann I believe that too, let's try to do one thing and do it good. Can we start from this? Raffo@bc23d5a

Sorry, I missed this one earlier. It looks like a proper direction, however I'm not sure about the implementation details. For example there are different error responses for `409` error and marathon returns different fields which we need to unmarshal. Also, maybe I would keep the errors definitions closer to the methods producing them.

**Comment by timoreimann** [comment]
I'm with @eicca here: Conceptually, we're on the same page, and implementation details need to be sorted out. Let's discuss in detail again once my PR is ready.

(Ideally, I'd like #103 to finish first as it changes quite a few places my PR would need to rely on.)

**Comment by Raffo** [comment]
@eicca No problem :-) That's why I said "start from here". I think it's the direction to go, we can sort the implementation details together.

**Comment by Raffo** [comment]
This has been stuck forever... any update?

**Comment by timoreimann** [comment]
I'm working on it in this very moment. Hope to send out a PR still within the week.

**Comment by timoreimann** [comment]
PR out.

Note that I haven't included the HTTP error codes despite this being one of my original motivations. The reason being that Marathon doesn't seem to come with as many different error types as I initially thought: I went through the entire [REST API](https://mesosphere.github.io/marathon/docs/generated/api.html) and tried to make sure that everything is covered. And even if something may have slipped through, any missing errors should be easy to add.

**Comment by Raffo** [comment]
I'll have a look at the PR and see if I can provide some valid comment. The Marathon API is indeed not good, the status code and the possibility for error handling are limited.

**Comment by timoreimann** [comment]
Will close the ticket as the PR has been merged. We can open a new ticket should the need to expose low level protocol information rise again.

Thanks to everyone involved!
